lib/FIR/flattened.h

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_FIR_FLATTENED_H_
+#define FORTRAN_FIR_FLATTENED_H_
+
+#include "common.h"
+#include "mixin.h"
+#include "../parser/parse-tree.h"
+#include <list>
+#include <map>
+#include <set>
+#include <vector>
+
+namespace Fortran::FIR {
+
+struct AnalysisData;
+
+namespace flat {
+
+// This is a flattened, linearized representation of the parse tree. It captures
+// the executable specification of the input program. The flattened IR can be
+// used to construct the Fortran IR.
+
+struct Op;
+struct LabelOp;
+struct GotoOp;
+struct ReturnOp;
+struct ConditionalGotoOp;
+struct SwitchIOOp;
+struct SwitchOp;
+struct ActionOp;
+struct BeginOp;
+struct EndOp;
+struct IndirectGotoOp;
+struct DoIncrementOp;
+struct DoCompareOp;
+
+using LabelRef = unsigned;
+constexpr LabelRef unspecifiedLabel{~0u};
+
+using Location = parser::CharBlock;
+struct LabelBuilder;
+
+// target for a control-flow edge
+struct LabelOp {
+  explicit LabelOp(LabelBuilder &builder);
+  LabelOp(const LabelOp &that);
+  LabelOp &operator=(const LabelOp &that);
+  void setReferenced() const;
+  bool isReferenced() const;
+  LabelRef get() const { return label_; }
+  operator LabelRef() const { return get(); }
+  void dump() const;
+
+private:
+  LabelBuilder &builder_;
+  LabelRef label_;
+};
+
+struct ArtificialJump {};
+constexpr ArtificialJump ARTIFICIAL{};
+
+// a source of an absolute control flow edge
+struct GotoOp
+  : public SumTypeCopyMixin<const parser::CycleStmt *, const parser::ExitStmt *,
+        const parser::GotoStmt *, ArtificialJump> {
+  template<typename A>
+  GotoOp(const A &stmt, LabelRef dest, const Location &source)
+    : SumTypeCopyMixin{&stmt}, target{dest}, source{source} {}
+  explicit GotoOp(LabelRef dest) : SumTypeCopyMixin{ARTIFICIAL}, target{dest} {}
```

Comment:
My assumption has been the once the FIR is built, the parse tree is no longer needed. But this seems to be saving pointers to parse tree nodes. Will these still be around when the FIR is completely built? Either way, it would be good to document what we need in terms of parse tree lifetime, though I don't have any idea where that should be.

Reply:
The flat form is an intermediate step along the lowering path. (This is just being moved and isn't anything new.) It does keep pointers into the parse tree (since it needs to reference the input/the tree to know how to proceed with further lowering), but the flat form is transitory and goes away.